client/src/main/java/io/pravega/client/stream/impl/ReaderGroupImpl.java

```diff
@@ -203,8 +214,41 @@ private Checkpoint completeCheckpoint(String checkpointName) {
 
     @Override
     public void resetReaderGroup(ReaderGroupConfig config) {
-        Map<SegmentWithRange, Long> segments = getSegmentsForStreams(controller, config);
-        synchronizer.updateStateUnconditionally(new ReaderGroupStateInit(config, segments, getEndSegmentsForStreams(config)));
+        while (true) {
+            synchronizer.fetchUpdates();
+            val state = synchronizer.getState();
+            // We only move into the block if the state transition has happened successfully.
+            if (stateTransition(state, new UpdatingConfig(true))) {
+                // Use the latest generation and reader group Id.
+                ReaderGroupConfig newConfig = config.toBuilder()
+                        .readerGroupId(state.getConfig().getReaderGroupId())
+                        .generation(state.getConfig().getGeneration()).build();
+                boolean success = Futures.getThrowingException(controller.updateReaderGroup(scope, groupName, newConfig));
+                if (success) {
+                    synchronizer.updateState((s, updates) -> {
+                        Map<SegmentWithRange, Long> segments = getSegmentsForStreams(controller, newConfig);
```

Comment:
getSegmentsForStreams is making an RPC to the controller. Please move this out of the updateState block.

Reply:
Moved out.